// eslint-plugin-testing-library/rules/consistent-data-testid.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var utils_1 = require("../utils");
var experimental_utils_1 = require("@typescript-eslint/experimental-utils");
var node_utils_1 = require("../node-utils");
exports.RULE_NAME = 'consistent-data-testid';
var FILENAME_PLACEHOLDER = '{fileName}';
exports.default = experimental_utils_1.ESLintUtils.RuleCreator(utils_1.getDocsUrl)({
    name: exports.RULE_NAME,
    meta: {
        type: 'suggestion',
        docs: {
            description: 'Ensures consistent usage of `data-testid`',
            category: 'Best Practices',
            recommended: false,
        },
        messages: {
            invalidTestId: '`{{attr}}` "{{value}}" should match `{{regex}}`',
        },
        fixable: null,
        schema: [
            {
                type: 'object',
                default: {},
                additionalProperties: false,
                required: ['testIdPattern'],
                properties: {
                    testIdPattern: {
                        type: 'string',
                    },
                    testIdAttribute: {
                        type: 'string',
                        default: 'data-testid',
                    },
                },
            },
        ],
    },
    defaultOptions: [
        {
            testIdPattern: '',
            testIdAttribute: 'data-testid',
        },
    ],
    create: function (context, _a) {
        var _b;
        var options = _a[0];
        var getFilename = context.getFilename;
        var testIdPattern = options.testIdPattern, attr = options.testIdAttribute;
        function getFileNameData() {
            var splitPath = getFilename().split('/');
            var fileNameWithExtension = splitPath.pop();
            var parent = splitPath.pop();
            var fileName = fileNameWithExtension.split('.').shift();
            return {
                fileName: fileName === 'index' ? parent : fileName,
            };
        }
        function getTestIdValidator(fileName) {
            return new RegExp(testIdPattern.replace(FILENAME_PLACEHOLDER, fileName));
        }
        return _b = {},
            _b["JSXIdentifier[name=" + attr + "]"] = function (node) {
                if (!node_utils_1.isJSXAttribute(node.parent) || !node_utils_1.isLiteral(node.parent.value)) {
                    return;
                }
                var value = node.parent.value.value;
                var fileName = getFileNameData().fileName;
                var regex = getTestIdValidator(fileName);
                if (value && typeof value === 'string' && !regex.test(value)) {
                    context.report({
                        node: node,
                        messageId: 'invalidTestId',
                        data: {
                            attr: attr,
                            value: value,
                            regex: regex,
                        },
                    });
                }
            },
            _b;
    },
});
